kong/clustering/version_negotiation/init.lua

```diff
@@ -0,0 +1,340 @@
+    end
+
+    local name = req_service.name
+
+    if type(req_service.versions) ~= "table" then
+      table.insert(services_rejected, {
+        name = name,
+        message = "invalid \"versions\" table.",
+      })
+      goto continue
+    end
+
+    local known_service = all_known_services[name]
+    if not known_service then
+      table.insert(services_rejected, {
+        name = name,
+        message = "unknown service.",
+      })
+      goto continue
+    end
+
+    local ok, service_response = cp_priority(name, req_service.versions, known_service)
+    if ok then
+      ngx_log(ngx_DEBUG, _log_prefix,
+        "accepted: \"" .. service_response.name ..
+        "\", version \"" .. service_response.version ..
+        "\": ".. service_response.message)
+      table.insert(services_accepted, service_response)
+    else
+
+      ngx_log(ngx_DEBUG, _log_prefix,
+        "rejected: \"" .. service_response.name ..
+        "\": " .. service_response.message)
+      table.insert(services_rejected, service_response)
+    end
+
+    ::continue::
+  end
+
+  return {
+    node = node_info(),
+    services_accepted = services_accepted,
+    services_rejected = services_rejected,
+  }
+end
+
+
+local function register_client(conf, client_node, services_accepted)
+  local services_map = {}
+  for _, service in ipairs(services_accepted) do
+    services_map[service.name] = service.version
+  end
+
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = client_node.id, }, {
+    last_seen = ngx.time(),
+    config_hash = "00000000000000000000000000000000",
+    hostname = client_node.hostname,
+    ip = ngx.var.remote_addr,
+    version = client_node.version,
+    sync_status = client_node.sync_status,
+    services_accepted = services_map,
+  }, { ttl = conf.cluster_data_plane_purge_delay })
+
+  if not ok then
+    ngx_log(ngx_ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return nil, err
+  end
+
+  return true
+end
+
+--- Handles a version negotiation request (CP side).
+--- Performs mTLS verification (as configured),
+--- validates request and Kong version compatibility,
+---
+function _M.serve_version_handshake(conf, cert_digest)
+  if KONG_VERSION == nil then
+    KONG_VERSION = kong.version
+  end
+
+  local ok, err = clustering_utils.validate_connection_certs(conf, cert_digest)
+  if not ok then
+    ngx_log(ngx_ERR, _log_prefix, err)
+    return ngx.exit(ngx.HTTP_CLOSE)
+  end
+
+  local body_in = cjson.decode(get_body())
+  if not body_in then
+    return response_err("Not valid JSON data")
+  end
+
+  ok, err = verify_request(body_in)
+  if not ok then
+    return response_err(err)
+  end
+
+  ok, err, body_in.node.sync_status = check_node_compatibility(body_in.node)
+  if not ok then
+    return response_err(err)
+  end
+
+  local body_out = do_negotiation(body_in)
+
+  ok, err = register_client(conf, body_in.node, body_out.services_accepted)
+  if not ok then
+    return response_err(err)
```

Comment:
Maybe this should be considered 500, rather than 400?

Reply:
done